core/src/main/java/io/questdb/cairo/wal/seq/TableSequencerImpl.java

```diff
@@ -0,0 +1,280 @@
+            doClose();
+            throw th;
+        }
+    }
+
+    @Override
+    public void close() {
+        schemaLock.writeLock().lock();
+        try {
+            doClose();
+        } finally {
+            schemaLock.writeLock().unlock();
+        }
+    }
+
+    @Override
+    public void copyMetadataTo(@NotNull SequencerMetadata copyTo) {
+        schemaLock.readLock().lock();
+        try {
+            copyTo.copyFrom(metadata);
+        } finally {
+            schemaLock.readLock().unlock();
+        }
+    }
+
+    @Override
+    public TableMetadataChangeLog getMetadataChangeLogCursor(long structureVersionLo) {
+        checkDistressed();
+        if (metadata.getStructureVersion() == structureVersionLo) {
+            // Nothing to do.
+            return EmptyOperationCursor.INSTANCE;
+        }
+        return tableTransactionLog.getTableMetadataChangeLog(structureVersionLo, alterCommandWalFormatter);
+    }
+
+    @Override
+    public TransactionLogCursor getTransactionLogCursor(long seqTxn) {
+        checkDistressed();
+        return tableTransactionLog.getCursor(seqTxn);
+    }
+
+    @Override
+    public int getNextWalId() {
+        return (int) walIdGenerator.getNextId();
+    }
+
+    @Override
+    public int getTableId() {
+        return metadata.getTableId();
+    }
+
+    @Override
+    public long nextStructureTxn(long expectedStructureVersion, TableMetadataChange change) {
+        // Writing to TableSequencer can happen from multiple threads, so we need to protect against concurrent writes.
+        schemaLock.writeLock().lock();
+        long txn;
+        try {
+            checkDistressed();
+            if (metadata.getStructureVersion() == expectedStructureVersion) {
+                long offset = tableTransactionLog.beginMetadataChangeEntry(expectedStructureVersion + 1, alterCommandWalFormatter, change);
+
+                applyToMetadata(change);
+                if (metadata.getStructureVersion() != expectedStructureVersion + 1) {
+                    throw CairoException.critical(0)
+                            .put("applying structure change to WAL table failed [table=").put(tableName)
+                            .put(", oldVersion: ").put(expectedStructureVersion)
+                            .put(", newVersion: ").put(metadata.getStructureVersion())
+                            .put(']');
+                }
+                txn = tableTransactionLog.endMetadataChangeEntry(expectedStructureVersion + 1, offset);
+            } else {
+                return NO_TXN;
+            }
+        } catch (Throwable th) {
+            distressed = true;
+            LOG.critical().$("could not apply structure change to WAL table sequencer [table=").$(tableName)
+                    .$(", error=").$(th.getMessage())
+                    .I$();
+            throw th;
+        } finally {
+            schemaLock.writeLock().unlock();
+        }
+        if (!metadata.isSuspended()) {
```

Comment:
Looks like a racy check.

Reply:
"a racy check" means that, at least, `SequencerMetadata#suspended` should have a `volatile` modifier. The current code is broken as it relies on compiler kindness and a strong hardware memory model.

Reply:
what i am saying is that it does not really matter when other threads will see the change (if they see it ever).
in practice we will hit a volatile or a fence somewhere in the code at some point anyway, but even if we do not it is fine.

Reply:
If it doesn't matter whether other threads see this change or not, let's remove this check. There is nothing preventing the compiler storing this value in a register until eternity. I don't see any convincing reason not to add a `volatile` modifier here. Do you see any?

Reply:
i think there will be a lot of unnecessary volatile reads here but added volatile to the suspended flag.